pkg/network/ebpf/c/protocols/kafka/helpers.h

```diff
@@ -0,0 +1,287 @@
+static __always_inline bool get_topic_offset_from_produce_request(const kafka_header_t *kafka_header, struct __sk_buff *skb, u32 *out_offset) {
+    const s16 api_version = kafka_header->api_version;
+    u32 offset = *out_offset;
+    if (api_version >= 3) {
+        READ_BIG_ENDIAN_WRAPPER(s16, transactional_id_size, skb, offset);
+        if (transactional_id_size > 0) {
+            offset += transactional_id_size;
+        }
+    }
+
+    READ_BIG_ENDIAN_WRAPPER(s16, acks, skb, offset);
+    if (acks > 1 || acks < -1) {
+        // The number of acknowledgments the producer requires the leader to have received before considering a request
+        // complete. Allowed values: 0 for no acknowledgments, 1 for only the leader and -1 for the full ISR.
+        return false;
+    }
+
+    READ_BIG_ENDIAN_WRAPPER(s32, timeout_ms, skb, offset);
+    if (timeout_ms < 0) {
+        // timeout_ms cannot be negative.
+        return false;
+    }
+
+    *out_offset = offset;
+    return true;
+}
+
+// Getting the offset the topic name in the fetch request.
+static __always_inline u32 get_topic_offset_from_fetch_request(const kafka_header_t *kafka_header, struct __sk_buff *skb) {
+    // replica_id => INT32
+    // max_wait_ms => INT32
+    // min_bytes => INT32
+    u32 offset = 3 * sizeof(s32);
+
+    if (kafka_header->api_version >= 3) {
+        // max_bytes => INT32
+        offset += sizeof(s32);
+        if (kafka_header->api_version >= 4) {
+            // isolation_level => INT8
+            offset += sizeof(s8);
+            if (kafka_header->api_version >= 7) {
+                // session_id => INT32
+                // session_epoch => INT32
+                offset += 2 * sizeof(s32);
+            }
+        }
+    }
+
+    return offset;
+}
+
+// Calls the relevant function, according to the api_key.
+static __always_inline bool is_kafka_request(const kafka_header_t *kafka_header, struct __sk_buff *skb, u32 offset) {
+    // Due to old-verifiers limitations, if the request is fetch or produce, we are calculating the offset of the topic
+    // name in the request, and then validate the topic. We have to have shared call for validate_first_topic_name
+    // as the function is huge, rather than call validate_first_topic_name for each api_key.
+    switch (kafka_header->api_key) {
+    case KAFKA_PRODUCE:
+        if (!get_topic_offset_from_produce_request(kafka_header, skb, &offset)) {
+            return false;
+        }
+        break;
+    case KAFKA_FETCH:
+        offset += get_topic_offset_from_fetch_request(kafka_header, skb);
+        break;
+    default:
+        return false;
+    }
+    return validate_first_topic_name(skb, offset);
+}
+
+// Checks if the packet represents kafka communication.
+static __always_inline bool is_kafka(struct __sk_buff *skb, skb_info_t *skb_info, const char* buf, __u32 buf_size) {
+    CHECK_PRELIMINARY_BUFFER_CONDITIONS(buf, buf_size, KAFKA_MIN_LENGTH);
```

Comment:
Do you really need buf or you can use skb to read the header?

Reply:
we already have it, why not using it?